File: Emotion.py
import os
import cv2
import numpy as np
import urllib.request
from keras.models import load_model
from colorama import Fore, Style

cascade_file_path = "corpdata/Fer2013-img/haarcascade_frontalface_default.xml"
if not os.path.exists(cascade_file_path):
    cv2_url = "https://github.com/opencv/opencv/blob/master/data/haarcascades/haarcascade_frontalface_default.xml"
    print(
        Fore.YELLOW
        + "Haar cascade classifier file not found. Downloading..."
        + Style.RESET_ALL
    )
    urllib.request.urlretrieve(cv2_url, cascade_file_path)
    print(
        Fore.GREEN
        + "Haar cascade classifier file downloaded successfully!"
        + Style.RESET_ALL
    )

emotion_model = load_model("face_emotion_model.h5")
face_cascade = cv2.CascadeClassifier(cascade_file_path)


def main():
    Capture = cv2.VideoCapture(0)
    while True:
        ret, frame = Capture.read()
        if not ret:
            break
        Grayscale = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        Faces = face_cascade.detectMultiScale(
            Grayscale, scaleFactor=1.1, minNeighbors=5, minSize=(30, 30)
        )
        for x, y, w, h in Faces:
            face_roi = Grayscale[y : y + h, x : x + w]
            face_roi = cv2.resize(face_roi, (48, 48))
            face_roi = face_roi.astype("float32") / 255.0
            face_roi = np.expand_dims(face_roi, axis=0)
            emotions = emotion_model.predict(face_roi)
            emotion_labels = [
                "Angry",
                "Disgust",
                "Fear",
                "Happy",
                "Sad",
                "Surprise",
                "Neutral",
            ]
        username = None
        predicted_emotion = emotion_labels[np.argmax(emotions)]
        for File in os.listdir("userNamedfaces"):
            img = cv2.imread(os.path.join("userNamedfaces", File), cv2.IMREAD_GRAYSCALE)
            img = cv2.resize(img, (48, 48))
            img = img.astype("float32") / 255.0
            img = np.expand_dims(img, axis=0)
            img_emotions = emotion_model.predict(img)
            img_predicted_emotion = emotion_labels[np.argmax(img_emotions)]
            if img_predicted_emotion == predicted_emotion:
                username = File.split(".")[0]
                break
        if username:
            print(
                Fore.CYAN
                + "Detected username: "
                + Fore.YELLOW
                + f"{username}"
                + Style.RESET_ALL
            )
            print(
                Fore.CYAN
                + "Predicted emotion: "
                + Fore.YELLOW
                + f"{predicted_emotion}"
                + Style.RESET_ALL
            )
        cv2.imshow("Emotion Detection", frame)
        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    Capture.release()
    cv2.destroyAllWindows()


main()

File: routes/exports/Emotion.js
import cv from "opencv4nodejs"; // yarn add opencv4nodejs
import tf from "@tensorflow/tfjs-node"; // yarn add @tensorflow/tfjs-node

const model = await tf.loadLayersModel("emotion_detection_model.h5");
const camera = new cv.VideoCapture(0);
camera.set(cv.CAP_PROP_FPS, 30);
cv.namedWindow("Emotion Detection", cv.WINDOW_NORMAL);

while (true) {
  const frame = camera.read();
  const grayFrame = frame.cvtColor(cv.COLOR_BGR2GRAY);
  const resizedFrame = grayFrame.resize(new cv.Size(28, 28));
  const normalizedFrame = resizedFrame.normalize(
    0,
    255,
    cv.NORM_MINMAX,
    cv.CV_32F
  );
  const tensorFrame = tf.tensor4d([normalizedFrame.getDataAsArray()]);
  const prediction = model.predict(tensorFrame);
  const predictionArray = prediction.arraySync()[0];
  const maxIndex = predictionArray.indexOf(Math.max(...predictionArray));
  const emotions = [
    "Angry",
    "Disgust",
    "Fear",
    "Happy",
    "Sad",
    "Surprise",
    "Neutral",
  ];
  const predictedEmotion = emotions[maxIndex];
  frame.putText(
    predictedEmotion,
    new cv.Point(10, 50),
    cv.FONT_HERSHEY_SIMPLEX,
    1,
    new cv.Vec(0, 0, 255),
    2
  );
  cv.imshow("Emotion Detection", frame);
  const key = cv.waitKey(100);
  if (key === 113) {
    break;
  }
}

camera.release();
cv.destroyAllWindows();
